### PR TITLE
add batch FD with OMP Offload

### PR DESCRIPTION
**Pull request by wjge**

### CODE REVIEW

**Comment by wjge** [comment]
I forgot to pass the number of functions which is still hard coded. I need to add that before the merge.

File: src/pb/FDoper.cc [hunk 1]
```diff
@@ -467,6 +467,108 @@ void FDoper<T>::del2_4th(GridFunc<T>& A, GridFunc<T>& B) const
 
     del2_4th_tm_.stop();
 }
+
+template <class T>
+void FDoper<T>::del2_4th(const Grid& Agrid, T* A, T* B, const size_t nfunc) const
```
Comment (jeanlucf22) [inline src/pb/FDoper.cc]: This function is a generalization of the previous one, with an extra parameter "nfunc". Is there a reason to keep the previous function, or should we now use that one with "nfunc=1" instead?

Comment (wjge) [reply src/pb/FDoper.cc]: This is a good idea. Let me replace the previous 3-D one in another PR. The only thing is now only the implementation of the 4th-order scheme is changed for OMP offload. What about the rest of the schemes?

Comment (wjge) [reply src/pb/FDoper.cc]: To 'replace it' I mean if the argument of apply() is a single function, the function will pass nfunc=1 and call the new 4-D implementation.

File: tests/testBatchLaph4.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            int iiy  = iy * grid.inc(1) + iix;
+            double y = grid.start(1) + iy * h[1];
+
+            for (int iz = nghosts; iz < endz; iz++)
+            {
+                double z = grid.start(2) + iz * h[2];
+
+                u1[iiy + iz]
+                    = sin(x * coeffx) + sin(y * coeffy) + sin(z * coeffz);
```
Comment (Rombur) [inline tests/testBatchLaph4.cc:55]: `sin` is C, use `std::sin`

File: tests/testBatchLaph4.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    const double h[3] = { ll / (double(ngpts[0])), ll / (double(ngpts[1])),
+        ll / (double(ngpts[2])) };
+    pb::PEenv mype_env(MPI_COMM_WORLD, ngpts[0], ngpts[1], ngpts[2]);
+    pb::Grid grid(origin, lattice, ngpts, mype_env, nghosts, 0);
+
+    pb::Laph4<double> lap(grid);
+
+    const size_t numgridfunc = 10;
+
+    double* arrayofgf1 = new double[numgridfunc * grid.sizeg()];
```
Comment (Rombur) [inline tests/testBatchLaph4.cc:24]: Why not use a `unique_ptr`?

File: tests/testBatchLaph4.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    // fill ghost values
+    gf1.trade_boundaries();
+
+    for (size_t inum = 0; inum < numgridfunc; inum++)
+    {
+        std::copy(gf1.uu(), gf1.uu() + grid.sizeg(),
+            arrayofgf1 + inum * grid.sizeg());
+    }
+
+    double* arrayofgf2 = new double[numgridfunc * grid.sizeg()];
```
Comment (Rombur) [inline tests/testBatchLaph4.cc:71]: `unique_ptr`?

File: tests/testBatchLaph4.cc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+TEST_CASE("Laplacian 4th order for a array of function", "[laph4_batch]")
+{
+    const double origin[3]  = { 0., 0., 0. };
+    const double ll         = 2.;
+    const double lattice[3] = { ll, ll, ll };
+    const unsigned ngpts[3] = { 32, 24, 20 };
+    const short nghosts     = 2;
+
+    const double h[3] = { ll / (double(ngpts[0])), ll / (double(ngpts[1])),
```
Comment (Rombur) [inline tests/testBatchLaph4.cc:15]: Use `static_cast`

**Comment by jeanlucf22** [comment]
@Rombur sorry, I guess I merged this PR too quickly... @wjge you can fix these things in your next PR

**Pull request merged**